extra_foam/gui/image_tool/reference_view.py

```diff
@@ -76,26 +102,106 @@ def _loadReference(self):
+        image = self._corrected.image
+        if image is not None:
+            self._writeTempReferenceFile(image)
+            if self._checkReferenceFile():
+                self._setReference(REFERENCE_FILE)
+                # Update label
+                text = "The current image has been applied as reference."
+                self._ctrl_widget.record_label.setText(text)
+
+    def _writeTempReferenceFile(self, image):
+        """Save the input image as a reference file (.npy).
+           Path is `REFERENCE_FILE`"""
+        if image is None:
+            return
+
+        os.makedirs(osp.dirname(REFERENCE_FILE), exist_ok=True)
+
+        try:
+            write_image(REFERENCE_FILE, image)
+        except ValueError as e:
+            logger.error(str(e))
+            # Remove previous reference file
+            if os.path.exists(REFERENCE_FILE):
+                os.remove(REFERENCE_FILE)
+
+    def _setReference(self, path):
+        """Sets the input image as reference"""
+        self._pub.set(path)
+        self._ctrl_widget.filepath_le.setText(path)
 
     @pyqtSlot()
     def _removeReference(self):
         """Remove the reference image."""
-        self._pub.set("")
-        self._ctrl_widget.filepath_le.setText("")
+        self._setReference("")
+        self._ctrl_widget.record_label.setText("")
+
+    @pyqtSlot(bool)
+    def _recordReference(self, is_recording):
+        """Start/stop the recording of the reference file.
+
+           When starting, a flag will be set for the `updateF()` to detect the
+           start of the image averaging of the received data.
+           Upon finishing, the recorded image is written as reference file.
+        """
+        self._is_recording = is_recording
+
+        # Log recording status
+        status = "started" if is_recording else "finished"
+        logger.info(f"Recording reference has {status}.")
+
+        if is_recording:
+            # Remove existing reference file
+            self._removeReference()
+        elif self._recorded_image is not None:
+            # Save recording to a temp file upon finishing
+            self._writeTempReferenceFile(self._recorded_image)
+            if self._checkReferenceFile():
+                self._setReference(REFERENCE_FILE)
+
+            # Reset widget and variables
+            self._recorded_image = None
+            self._count = 0
+
+            # Update label
+            text = ("The average of recorded trains "
+                    "has been applied as reference.")
+            self._ctrl_widget.record_label.setText(text)
+
+    @pyqtSlot()
+    def _saveReference(self):
+        """'Save' the reference file on the user supplied path by conveniently
+            copying the existing file to the desired path."""
+        if not self._checkReferenceFile():
+            return
+
+        filepath = QFileDialog.getSaveFileName(
+            caption="Save image",
+            directory=osp.expanduser("~"),
```

Comment:
may be a good idea to have `last_saved_location` from the filepath. and the next time if user wants to save again, use that as the starting directory in the FileDialog.

Reply:
I'd like that idea :D prolly it would be better to implement it in a more general manner. Would think more about this, perhaps on the next MRs.

Reply:
yes makes sense. Thanks